lib/runtime/wazero/instance.go

```diff
@@ -0,0 +1,828 @@
+		WithFunc(ext_storage_read_version_1).
+		Export("ext_storage_read_version_1").
+		NewFunctionBuilder().
+		WithFunc(ext_storage_root_version_1).
+		Export("ext_storage_root_version_1").
+		NewFunctionBuilder().
+		WithFunc(ext_storage_root_version_2).
+		Export("ext_storage_root_version_2").
+		NewFunctionBuilder().
+		WithFunc(ext_storage_set_version_1).
+		Export("ext_storage_set_version_1").
+		NewFunctionBuilder().
+		WithFunc(ext_storage_start_transaction_version_1).
+		Export("ext_storage_start_transaction_version_1").
+		NewFunctionBuilder().
+		WithFunc(ext_storage_rollback_transaction_version_1).
+		Export("ext_storage_rollback_transaction_version_1").
+		NewFunctionBuilder().
+		WithFunc(ext_storage_commit_transaction_version_1).
+		Export("ext_storage_commit_transaction_version_1").
+		Instantiate(ctx)
+
+	if err != nil {
+		panic(err)
+	}
+
+	code, err = decompressWasm(code)
+	if err != nil {
+		return nil, err
+	}
+
+	mod, err := rt.Instantiate(ctx, code)
+	if err != nil {
+		return nil, err
+	}
+
+	global := mod.ExportedGlobal("__heap_base")
+	if global == nil {
+		return nil, fmt.Errorf("wazero error: nil global for __heap_base")
+	}
+
+	hb := api.DecodeU32(global.Get())
+	// hb = runtime.DefaultHeapBase
+
+	mem := mod.Memory()
+	if mem == nil {
+		return nil, fmt.Errorf("wazero error: nil memory for module")
+	}
+
+	allocator := runtime.NewAllocator(mem, hb)
+
+	return &Instance{
+		Runtime: rt,
+		Context: &runtime.Context{
+			Storage:         cfg.Storage,
+			Allocator:       allocator,
+			Keystore:        cfg.Keystore,
+			Validator:       cfg.Role == common.AuthorityRole,
+			NodeStorage:     cfg.NodeStorage,
+			Network:         cfg.Network,
+			Transaction:     cfg.Transaction,
+			SigVerifier:     crypto.NewSignatureVerifier(logger),
+			OffchainHTTPSet: offchain.NewHTTPSet(),
+		},
+		Module: mod,
+	}, nil
+}
+
+var ErrExportFunctionNotFound = errors.New("export function not found")
+
+func (i *Instance) Exec(function string, data []byte) (result []byte, err error) {
+	i.Lock()
+	defer i.Unlock()
+
+	dataLength := uint32(len(data))
+	inputPtr, err := i.Context.Allocator.Allocate(dataLength)
+	if err != nil {
+		return nil, fmt.Errorf("allocating input memory: %w", err)
+	}
+
+	defer i.Context.Allocator.Clear()
+
+	// Store the data into memory
+	mem := i.Module.Memory()
+	if mem == nil {
+		panic("nil memory")
+	}
+	ok := mem.Write(inputPtr, data)
+	if !ok {
+		panic("wtf?")
+	}
+
+	runtimeFunc := i.Module.ExportedFunction(function)
+	if runtimeFunc == nil {
+		return nil, fmt.Errorf("%w: %s", ErrExportFunctionNotFound, function)
+	}
+
+	ctx := context.WithValue(context.Background(), runtimeContextKey, i.Context)
+
+	values, err := runtimeFunc.Call(ctx, api.EncodeI32(int32(inputPtr)), api.EncodeI32(int32(dataLength)))
+	if err != nil {
+		return nil, fmt.Errorf("running runtime function: %w", err)
+	}
+	if len(values) == 0 {
+		return nil, fmt.Errorf("no returned values from runtime function: %s", function)
+	}
```

Comment:
is expected for every runtime function to return a value?

Reply:
In our case yes.  Any functions that we call exported by the Polkadot/Westend runtimes return something.